wallet/src/actors/app/methods.rs

```diff
@@ -818,4 +829,141 @@ impl App {
+            "password",
+            "Password must be at least 8 characters",
+        ))
+    } else {
+        Ok(password)
+    };
+    let overwrite = overwrite.unwrap_or(false);
+    app::combine_field_errors(source, password, move |seed_source, password| Validated {
+        description,
+        name,
+        overwrite,
+        password,
+        seed_source,
+    })
+    .map_err(validation_error)
+}
+
+/// Validate XPRV
+pub fn validate_xprv(
+    seed_data: types::Password,
+    backup_password: Option<types::Password>,
+) -> Result<types::SeedSource> {
+    let backup_password = backup_password.ok_or_else(|| {
+        validation_error(app::field_error(
+            "backup_password",
+            "Backup password not found for XPRV key",
+        ))
+    })?;
+    let seed_data_string = std::str::from_utf8(seed_data.as_ref()).map_err(|_| {
+        validation_error(app::field_error(
+            "seed_data",
+            "Could not convert seed data to XPRV string",
+        ))
+    })?;
+    let (hrp, ciphertext) = bech32::decode(seed_data_string).map_err(|_| {
+        validation_error(app::field_error(
+            "seed_data",
+            "Could not decode bench32 key",
+        ))
+    })?;
+
+    let decrypted_key_string = bech32::FromBase32::from_base32(&ciphertext)
+        .map_err(|_| {
+            validation_error(app::field_error(
+                "seed_data",
+                "Could not convert bech 32 decoded key to u8 array",
+            ))
+        })
+        .and_then(|res: Vec<u8>| {
+            crypto::decrypt_cbc(&res, backup_password.as_ref()).map_err(|_| {
+                validation_error(app::field_error("seed_data", "Could not decrypt seed data"))
+            })
+        })
+        .and_then(|decrypted: Vec<u8>| {
+            std::str::from_utf8(&decrypted)
+                .map(|str| str.to_string())
+                .map_err(|_| {
+                    validation_error(app::field_error("seed_data", "Could not decrypt seed data"))
+                })
+        })?;
+    match hrp.as_str() {
+        "xprv" => Ok(types::SeedSource::Xprv(decrypted_key_string.into())),
+        "xprvdouble" => {
+            let ocurrences: Vec<(usize, &str)> =
+                decrypted_key_string.match_indices("xprv").collect();
+            // xprvDouble should only have 2 ocurrences
+            if ocurrences.len() != 2 {
+                return Err(validation_error(app::field_error(
+                    "seed_data",
+                    "Invalid number of XPRV keys found for xprvDouble type",
+                )));
+            }
+            let (internal, external) = decrypted_key_string.split_at(ocurrences[1].0);
+
+            Ok(types::SeedSource::XprvDouble((
+                internal.into(),
+                external.into(),
+            )))
+        }
+        _ => Err(validation_error(app::field_error(
+            "seed_data",
+            "Invalid seed data prefix",
+        ))),
```

Comment:
All this logic inside the `xprvdouble` branch should be a separate function, and it should be tested. Something like `split_xprvdouble`.

Reply:
I see that it's already tested in `test_valid_xprv_double_decryption`, so doesn't need many tests. Maybe check that it doesn't panic with some strange inputs like

```
""
"xprv"
"xprvxprv"
"xprv1xprv2"
```

But it should work fine :+1:

Reply:
The issue for such a test is that I need a valid key, a valid encryption and a valid bech32 encoding with something different than "xprv" or "xprvdouble". I will try to make this work but I might have to obtain a test vector manually for that

Reply:
I forgot to mention that this happends because there is a bech32 decode that returns a different error if the decoding went wrong